File: 150to300/234_palindromeLinkedList.java
/**
1. 首先linkedlist populate所有值，并且一一存在 一个 新的linkedlist中。
2. one pass 新的linkedlist的首尾，如果有一刻首位不等那就是false 不然是 true

 */
class Solution {
    private LinkedList<Integer> list = new LinkedList<>();
    
    public boolean isPalindrome(ListNode head) {
        if (head == null) return true;
        while (head != null){
            list.add(head.val);
            head = head.next;
        }
        
        if (list.size() == 1) return true; 
        
        boolean ans = isPali(list);
        return ans;
    }
    
    private boolean isPali(LinkedList<Integer> list){
        boolean ans = true;
        /*
        需要预防奇偶问题， [1 0 1] 这个case， 如果用 while (!list.isEmpty())来判断 会在第二个循环的时候导致pollLast null 
        pointer。 所以其实如果list里还剩下一个的时候 就可以停止了，因为1个数也是pali
        */
        while (list.size() > 1) {
            int first = list.pollFirst();
            int last = list.pollLast();
            
            if (first != last){
                return false;
            }
        }
        return ans;
    }
}

File: 150to300/234_palindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
'''
Example 1:

Input: 1->2
Output: false
Example 2:

Input: 1->2->2->1
Output: true

1. copy everything into an array
2. two pointer check

'''
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head:
            return True
        
        array = []
        while(head):
            array.append(head.val)
            head = head.next
        return self.isPali(array)
            
    def isPali(self, array):
        start = 0
        end = len(array) - 1
        while(start < end):
            if array[start] != array[end]:
                return False
            start +=1
            end -=1
        return True
        
        